astropy/cosmology/parameter.py

```diff
@@ -0,0 +1,261 @@
+
+        # validate value, generally setting units if present
+        value = self.validate(cosmology, value)
+        setattr(cosmology, self._attr_name_private, value)
+
+    # -------------------------------------------
+    # validate value
+
+    @property
+    def fvalidate(self):
+        """Function to validate a potential value of this Parameter.."""
+        return self._fvalidate
+
+    def validator(self, fvalidate):
+        """Make new Parameter with custom ``fvalidate``.
+
+        Note: ``Parameter.fvalidator`` must be the top-most descriptor decorator.
+
+        Parameters
+        ----------
+        fvalidate : callable[[type, type, Any], Any]
+
+        Returns
+        -------
+        `~astropy.cosmology.Parameter`
+            Copy of this Parameter but with custom ``fvalidate``.
+        """
+        desc = type(self)(fvalidate=fvalidate,
+                          doc=self.__doc__, fmt=self.format_spec,
+                          unit=self.unit, equivalencies=self.equivalencies,
+                          derived=self.derived)
+        return desc
+
+    def validate(self, cosmology, value):
+        """Run the validator on this Parameter.
+
+        Parameters
+        ----------
+        cosmology : `~astropy.cosmology.Cosmology` instance
+        value : Any
+            The object to validate.
+
+        Returns
+        -------
+        Any
+            The output of calling ``fvalidate(cosmology, self, value)``
+            (yes, that parameter order).
+        """
+        return self.fvalidate(cosmology, self, value)
+
+    @classmethod
+    def register_validator(cls, key, fvalidate=None):
+        """Decorator to register a new kind of validator function.
+
+        Parameters
+        ----------
+        key : str
+        fvalidate : callable[[object, object, Any], Any] or None, optional
+            Value validation function.
+
+        Returns
+        -------
+        ``validator`` or callable[``validator``]
+            if validator is None returns a function that takes and registers a
+            validator. This allows ``register_validator`` to be used as a
+            decorator.
+        """
+        if key in cls._registry_validators:
+            raise KeyError(f"validator {key!r} already registered with Parameter.")
+
+        # fvalidate directly passed
+        if fvalidate is not None:
+            cls._registry_validators[key] = fvalidate
+            return fvalidate
+
+        # for use as a decorator
+        def register(fvalidate):
+            """Register validator function.
+
+            Parameters
+            ----------
+            fvalidate : callable[[object, object, Any], Any]
+                Validation function.
+
+            Returns
+            -------
+            ``validator``
+            """
+            cls._registry_validators[key] = fvalidate
+            return fvalidate
+
+        return register
+
+    @classproperty
```

Comment:
I'd not add this - it is useful for developers only and those can be pointed to the private attribute.

Reply:
Removed.